### PR TITLE
[Merged by Bors] - Maintain attestations that reference unknown blocks

### PR DESCRIPTION
**Pull request by divagant-martian**
## Issue Addressed

#635 

## Proposed Changes
- Keep attestations that reference a block we have not seen for 30secs before being re processed
- If we do import the block before that time elapses, it is reprocessed in that moment
- The first time it fails, do nothing wrt to gossipsub propagation or peer downscoring. If after being re processed it fails, downscore with a `LowToleranceError` and ignore the message.

### CODE REVIEW

**Comment by terencechain** [comment]
Hi lighthouse team, we are looking to merge a prysm bug fix:

https://github.com/prysmaticlabs/prysm/pull/8827

Background, currently prysm validators always wait 1/3 of a slot before sending out attestation. We are a bit hesitant to merge the "timely" fix because @potuz pointed out that lighthouse doesn't hold attestations for unknown block. By merging the fix, it may impact large stakers where they often propose and attest in the same slot. I'm wondering if you guys have any thoughts on whether we can merge our timely fix or a timeline for #2319? Thanks so much!

File: beacon_node/network/src/beacon_processor/tests.rs
Comment (michaelsproul) [inline beacon_node/network/src/beacon_processor/tests.rs]: Should be uncommented

File: consensus/types/src/eth_spec.rs
Comment (michaelsproul) [inline consensus/types/src/eth_spec.rs]: These `Unpin` bounds don't seem to be required for anything, can we delete them?

File: beacon_node/network/src/metrics.rs
Comment (michaelsproul) [inline beacon_node/network/src/metrics.rs]:
```suggestion
        "Total number of unaggregated attestations that referenced an unknown block and were re-queued."
```

File: beacon_node/network/src/metrics.rs
Comment (michaelsproul) [inline beacon_node/network/src/metrics.rs]:
```suggestion
        "Total number of aggregated attestations that referenced an unknown block and were re-queued."
```

File: common/slot_clock/src/lib.rs
Comment (michaelsproul) [inline common/slot_clock/src/lib.rs]: I tried to get rid of this bound, but got bogged down in `async` weirdness so I'm content with it staying.

As far as I understand it's required by the `Stream` implementation for `ReprocessQueue`, but I'm a bit unsure how to describe exactly _why_ :sweat_smile:

File: beacon_node/network/src/beacon_processor/mod.rs
Comment (michaelsproul) [inline beacon_node/network/src/beacon_processor/mod.rs]:
```suggestion
    UnknownBlockAggregate {
```

File: beacon_node/network/src/beacon_processor/mod.rs
Comment (michaelsproul) [inline beacon_node/network/src/beacon_processor/mod.rs]:
```suggestion
    /// A work event that was queued for re-processing has become ready.
```

**Review by michaelsproul — CHANGES_REQUESTED** [review]
This looks really good based on my first read. I'd like @paulhauner's input still because he's the original author of the delay queue, but more importantly because we need to make a call about [signature verification of early attestations](https://hackmd.io/@CzkIH-W1R42R-S4vdcDJeQ/rJmmumLPu).

As far as I can tell we have two options:

1. Don't verify signatures on early attestations, and rely on the combination of a queue size bound & peer reputation to prevent DoS attacks (this is what the PR does now). If someone fills our queue with junk => they get banned.
2. Verify signatures on early attestations based on the target using `obtain_indexed_attestation_and_committees_per_slot` (similar to what `process_slash_info` does). There are a few extra complications in this approach:
    i. The case where the target is unknown because it's the first slot of the epoch, or because it points to a deep fork that our node doesn't know about. We could _try_ to achieve slightly better coverage by trying the head's shuffling when we have no other shuffling on hand, on the assumption that this gets us out of trouble on the 0th slot of the epoch on the canonical chain. It's still not perfect though, and we'd have to drop some attestations.
    ii. The logic in `attestation_verification.rs` doesn't currently allow us to encode the case where an attestation's _signature_ has been checked, but nothing else about it has. This would be necessary if we wanted to avoid re-verifying signatures on delayed attestations. This re-verification could prove a substantial burden if attestations increasingly arrive early due to eager publishers.

Perhaps it's worth leaning towards (1) for now, and revisiting (2) or similar if an effective attack is demonstrated against (1).

File: consensus/types/src/eth_spec.rs
Comment (divagant-martian) [reply consensus/types/src/eth_spec.rs]: They are required to be included in the Stream impl

File: beacon_node/network/src/beacon_processor/mod.rs
Comment (divagant-martian) [reply beacon_node/network/src/beacon_processor/mod.rs]: oh damn it looks like I have a misspell tick

File: consensus/types/src/eth_spec.rs
Comment (michaelsproul) [reply consensus/types/src/eth_spec.rs]: I tried deleting them and the stream impl still compiled

Comment (divagant-martian) [reply consensus/types/src/eth_spec.rs]: oh you are right. Unexpected (and removed) :+1:

File: common/slot_clock/src/lib.rs
Comment (divagant-martian) [reply common/slot_clock/src/lib.rs]: since only the clock was left I just passed it as reference and with that, byebye Unpin

Comment (michaelsproul) [reply common/slot_clock/src/lib.rs]: nice!

**Review by michaelsproul — APPROVED** [review]
I'm happy to merge this, modulo a decision in favour of (1) above.

**Comment by paulhauner** [comment]
> I'm happy to merge this, modulo a decision in favour of (1) above.

I'm also in favor of (1) with a well-bounded queue.

File: beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs
Comment (paulhauner) [inline beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs]: Super minor lint, but perhaps `fn beacon_block_root` would disambiguate between that and `aggregate.tree_hash_root()`?

File: beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs
Comment (paulhauner) [inline beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs]: As per above "super minor lint".

**Comment by paulhauner** [comment]
FYI I am in the middle of a review of this. I'll finish it over the weekend or on Monday.

Also, FYI, there's some merge conflicts now.

**Comment by divagant-martian** [comment]
> FYI I am in the middle of a review of this. I'll finish it over the weekend or on Monday.
> 
> Also, FYI, there's some merge conflicts now.

It was just a `debug!` that became a `trace!` in unstable. Solved now

File: beacon_node/network/src/beacon_processor/worker/gossip_methods.rs
Comment (paulhauner) [inline beacon_node/network/src/beacon_processor/worker/gossip_methods.rs:466]: I notice that we only push this message when we get a gossip block. Perhaps we should also be calling it when we process a `Working::process_rpc_block` (i.e, a blocks-by-root response)?

I imagine that notifying on blocks-by-root would improve this flow:

1. Attestation to unknown block arrives
2. The sync service requests the block from the peer
3. The peer gives us the block
4. The block arrives via gossip and is [rejected](https://github.com/sigp/lighthouse/blob/3b600acdc5bf9726367c18277a22486573b8b457/beacon_node/network/src/beacon_processor/worker/gossip_methods.rs#L285) since we already know about it.
5. The attestation expires after 30s and is the successfully imported.

File: beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs
Comment (paulhauner) [inline beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs]:
```suggestion
                        "Aggregate attestation delay queue is full";
```

Just to help us disambiguate between this and the unaggregate errors.

File: beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs
Comment (paulhauner) [inline beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs]:
```suggestion
pub const QUEUED_ATTESTATION_DELAY: Duration = Duration::from_secs(12);
```

I'm not *sure* about this, but I'm thinking that one (mainnet) slot would allow us to have a smaller queue and encounter the scenario where an attestation "expires" less (i.e., the slot becomes to old to import).

Do you have any objections here?

File: beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs
Comment (paulhauner) [reply beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs]: Included in https://github.com/paulhauner/lighthouse/commit/8fae9050e4d34859d793e74adee399033282afea

File: beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs
Comment (paulhauner) [reply beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs]: Included in https://github.com/paulhauner/lighthouse/commit/8fae9050e4d34859d793e74adee399033282afea

File: beacon_node/network/src/beacon_processor/worker/gossip_methods.rs
Comment (paulhauner) [reply beacon_node/network/src/beacon_processor/worker/gossip_methods.rs:466]: I implemented this here: https://github.com/paulhauner/lighthouse/commit/6744e2a7df3e0d6227cf4cb77478719f7616b8cb

**Review by paulhauner — CHANGES_REQUESTED** [review]
Heyo! This looks really good, I can't fault it structurally. I had a couple of suggestions, which I've either added to [my branch](https://github.com/paulhauner/lighthouse/commits/diva-attn-queueing) or I've made as single-line Github suggestions.

Feel free to merge or cherry-pick from my branch, as you see fit. There's some cases where I've made a comment on this PR and then implemented my suggestion on my branch; in such cases I've also provided a link to the specific commit.

For convenience, here's a full diff of my suggested changes: https://github.com/paulhauner/lighthouse/compare/e762f175b753fd61469a532ff139844265a76998..dca09d012155f6224291b0a56f009aa6a61ef0a2

I hope that providing actual commits instead of just PR comments is helpful. I've found that it's easier to communicate these things in Rust rather than English, plus it helps me filter out some dumb suggestions before they waste your time :sweat_smile:

File: beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs
Comment (michaelsproul) [inline beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs:49]: @dapplion astutely pointed out that this value may be too low if a large percentage of validators attest early. Assuming 250k validators, that's 7.8k per slot. We could set the max capacity to 8192 to give us a bit of headroom. What do you think @paulhauner, @divagant-martian?

File: beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs
Comment (divagant-martian) [reply beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs]: I thought a lot about this, without a clear solution. If I remember correctly the attn expires after an epoch and the 30secs are from a _reasonable time to complete the rpc request_. I'm fine changing it, since if it reaches the deadline, it means the block never arrived anyway

File: beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs
Comment (paulhauner) [reply beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs:49]: I agree. I also expected this to be a bit small but was planning to just run it and see what happens whilst it's in `unstable`.

That being said, I think 8,192 is a better starting point.

File: beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs
Comment (paulhauner) [reply beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs]: Hmm, yeah there's no clear answer. Regarding a "reasonable time", I would think that time needs to be less than 12s since that's our synchronicity assumption in Eth2; if peers can't dish out blocks before their slots are finished then it becomes basically impossible to build a common chain.

For this reason, I'd lean towards the 12s. However, there might still be something I'm missing.

File: beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs
Comment (dapplion) [reply beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs:49]: Why not compute these numbers on start-up depending on the preset constants?

Comment (paulhauner) [reply beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs:49]: > Why not compute these numbers on start-up depending on the preset constants?

I'm not sure I understand. Are you saying to compute it based on the current validator count?

I'm hesitant to over-think this, I'm keen to get it onto our staging infra, monitor it on Grafana and then start tweaking. The worst case is that the queue fills up and we fall back to the same behaviour as before this PR.

**Comment by paulhauner** [comment]
Hey @divagant-martian, just checking if this is still "waiting on author" or if it's "ready for review" :pray:

**Comment by paulhauner** [comment]
I've marked this as ready-for-review as per offline discussions with @divagant-martian. Feel free to remove the label if I misread Diva :slightly_smiling_face:

**Comment by divagant-martian** [comment]
Hey! I'm back. It's ready for review. I included all your proposed changes via merge. It was missing a decision about queue length but since that's solved you can give it a check again :)

**Comment by michaelsproul** [comment]
I think this needs to merge `unstable` again to bring in the Altair changes. Even though there are no textual conflicts I suspect there will be semantic ones

**Comment by divagant-martian** [comment]
Merge went out fine. What kind of conflicts to you think we are not noticing? Also it seems to be there is now a cargo audit issue that I'm checking

**Comment by michaelsproul** [comment]
Ah cool. I thought there might be some `state.slot`s that needed to change to `state.slot()` or something, nothing major

**Review by paulhauner — APPROVED** [review]
This is great, happy to merge once `cargo-audit` is appeased. I think @michaelsproul is working on that :relaxed:

**Pull request closed without merging**
